### PR TITLE
Update Install and Render handlers to tolerate v3 packages with optional marathon templates

### LINKED ISSUE #410: Correctly handle packages with no marathon template
[In `PackageInstallHandler`](https://github.com/dcos/cosmos/blob/c5546e5653535552e8b1e9ede10176d07843d1d9/cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageInstallHandler.scala#L81), the template is treated as empty if not specified, which is not correct:

```scala
    val marathonTemplate = pkg.marathon match {
      case Some(marathon) =>
        val bytes = ByteBuffers.getBytes(marathon.v2AppMustacheTemplate)
        new String(bytes, Charsets.Utf8)
      case _ => ""
    }
```

Determine what the correct behavior should be and update all usages of the template accordingly.

### LINKED ISSUE #417: Investigate the render endpoint with latest changes.
- [x] Test that the render endpoint works with version 2 universe which requires a Marathon template
- [ ] ~~Test that the render endpoint works with version 3 universe without a Marathon template.~~

### LINKED ISSUE #419: install (with a v2 request) of a package without marathon fails
screen shot 2016-06-27 at 5 14 06 pm
screen shot 2016-06-27 at 5 13 31 pm

### PR DESCRIPTION
**Pull request by BenWhitehead**
Depends on #421 
Fixes #419 
Fixes #410 
Fixes #417 

When reviewing, please only look at the contents of the second commit.

### CODE REVIEW

**Comment by BenWhitehead** [comment]
@tamarrow @jsancio Please review

File: cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala [hunk 1]
```diff
@@ -148,6 +154,60 @@ final class PackageInstallIntegrationSpec extends FreeSpec with BeforeAndAfterAl
...
+      installPackageAndAssert(
+        InstallRequest("enterprise-security-cli"),
+        expectedResult = Failure(Status.BadRequest, errorResponse),
+        preInstallState = Anything,
+        postInstallState = Unchanged
+      )
+    }
+
+    "will succeed if attempting to install a service for a v2 response with no marathon template" in {
+      import com.mesosphere.universe.v3.model._
```
Comment (tamarrow) [inline cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala]: why did you inline the import?

File: cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala [hunk 2]
```diff
@@ -148,6 +154,60 @@ final class PackageInstallIntegrationSpec extends FreeSpec with BeforeAndAfterAl
...
+        .buildPost(Buf.Utf8(installRequest.asJson.noSpaces))
+
+      val response = CosmosClient(request)
+
+      assertResult(Status.Ok)(response.status)
+      assertResult(MediaTypes.V2InstallResponse.show)(response.contentType.get)
+      val Xor.Right(actualBody) = decode[rpc.v2.model.InstallResponse](response.contentString)
+      assertResult(expectedBody)(actualBody)
+    }
+
```
Comment (tamarrow) [inline cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala]: thanks for adding tests for this!

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/Encoders.scala [hunk 3]
```diff
@@ -222,6 +223,8 @@ object Encoders {
           s"Unsupported redirect scheme - supported: $supportedMsg"
       }
     case ConversionFailure(message) => message
+    case ServiceMarathonTemplateNotFound(name, PackageDefinition.Version(version)) =>
+      s"Package with name: [$name] version: [$version] does not have a marathon template defined and can not be rendered"
```
Comment (tamarrow) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/Encoders.scala]: remove 'with name'

Comment (tamarrow) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/circe/Encoders.scala]: Marathon should be capitalized

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageInstallHandler.scala [hunk 4]
```diff
@@ -41,16 +41,28 @@ private[cosmos] final class PackageInstallHandler(
...
+                  cli = pkg.resource.flatMap(_.cli)
+                )
+              }
+          case None =>
+            Future {
+              rpc.v2.model.InstallResponse(
+                packageName = pkg.name,
+                packageVersion = pkg.version,
+                postInstallNotes = pkg.postInstallNotes,
+                cli = pkg.resource.flatMap(_.cli)
```
Comment (tamarrow) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageInstallHandler.scala]: lets explicitly add `appId = None`

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v2/circe/MediaTypedEncoders.scala [hunk 5]
```diff
@@ -32,9 +32,9 @@ object MediaTypedEncoders {
         mediaType = MediaTypes.V2InstallResponse
       ),
       MediaTypedEncoder(
-        encoder = rpc.v1.circe.Encoders.encodeInstallResponse.contramap(
-          converter.Response.v2InstallResponseToV1InstallResponse
-        ),
+        encoder = rpc.v1.circe.Encoders.encodeInstallResponse.contramap { (x: rpc.v2.model.InstallResponse) =>
+          converter.Response.v2InstallResponseToV1InstallResponse(x).get()
```
Comment (tamarrow) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v2/circe/MediaTypedEncoders.scala]: if this throws, does the user still get the appropriate response headers?

File: cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala (hunk 1 above)
Comment (BenWhitehead) [inline cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageInstallIntegrationSpec.scala]: So I didn't have to do the long form names when building the cli piece of the expectedBody.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v2/circe/MediaTypedEncoders.scala (hunk 5 above)
Comment (BenWhitehead) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v2/circe/MediaTypedEncoders.scala]: This this throw the user will get an error of type `ServiceMarathonTemplateNotFound` response back.

File: cosmos-server/src/it/scala/com/mesosphere/cosmos/handler/PackageRenderHandlerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      val response = CosmosClient(request)
+
+      assertResult(Status.BadRequest)(response.status)
+      assertResult(MediaTypes.ErrorResponse.show)(response.contentType.get)
+      val Xor.Right(actualBody) = decode[ErrorResponse](response.contentString)
+      assertResult(expectedBody)(actualBody)
+    }
+
+  }
+}
```
Comment (jsancio) [inline cosmos-server/src/it/scala/com/mesosphere/cosmos/handler/PackageRenderHandlerSpec.scala]: 👍

**Comment by jsancio** [comment]
LGTM! :shipit:

**Pull request merged**